### DISCUSSION TITLE
fastapi0.87.0 - ModuleNotFoundError: No module named 'httpx' while import TestClient from fastapi.testclient

### DISCUSSION THREAD

**Discussion by anoopks12** [fastapi/fastapi#6195] [Questions] [answered]
### First Check

- [X] I added a very descriptive title to this issue.
- [X] I used the GitHub search to find a similar issue and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/samuelcolvin/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
pip insall fastapi

>>> print(fastapi.__version__)
0.87.0

>>> from fastapi.testclient import TestClient
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "C:\ProgramData\Anaconda3\lib\site-packages\fastapi\testclient.py", line 1, in <module>
    from starlette.testclient import TestClient as TestClient  # noqa
  File "C:\ProgramData\Anaconda3\lib\site-packages\starlette\testclient.py", line 15, in <module>
    import httpx
ModuleNotFoundError: No module named 'httpx'
```

### Description

Getting error ModuleNotFoundError: No module named 'httpx' while import TestClient from fastapi.testclient.
It looks like an issue specific to the version 0.87.0 (which is released on 13-Nov-2022)
Its working in version 0.86.0

### Operating System

Windows

### FastAPI Version

0.87.0

### Python Version

3.8

**Comment by jgould22** [comment]
Yes, starlette has changed the TestClient to use HTTPX instead of requests and 0.87.0 upgrades starlette.

https://github.com/tiangolo/fastapi/releases/tag/0.87.0

The docs have been updated here https://fastapi.tiangolo.com/tutorial/testing/ to reflect the change.

**Comment by iudeen** [comment, marked as answer]
Just do `pip install httpx` or `pip install fastapi[all] —upgrade`

_If the solution above works for you, please close the issue._ 😊

**Comment by tiangolo** [comment]
Yep, install HTTPX, that's it 🚀

**Comment by mezhaka** [comment]
@tiangolo I think there's something wrong with installing `httpx`--we are pulling a transitive dependency of starlette directly: FastAPI -> Starlette -> httpx. Next time starlette migrates from httpx to httpy, this question will pop again.

Solution:
The httpx is included if one does `pip install starlette[full]`. Maybe FastAPI can pull `starlette[full]` if I do `pip install fastapi[all]`? This would allow starlette to manage its dependencies.

For the conda users: there is a `starlette-full` conda package that pulls httpx automatically, so just put `starlett-full` in your environment along with fastapi. It is better to have a first order transitive dependency in you environment FastAPI -> Starlette, rather the second order: FastAPI -> Starlette -> httpx.

@tiangolo I noticed there is no `fastapi-all` package on Anaconda? Can we create that as well? Then installing fastapi-all can pull starlette-full and this transitive dependency on httpx would be hidden. Who is maintaining current Anaconda package?

**Comment by ShantanuKumar** [comment]
yeah we use tool like [pants](https://www.pantsbuild.org/) which relies on transitive dependencies being explicit and hence it complains that it can't find `httpx`.  So installing manually transitive dependencies isn't great.

**Comment by iudeen** [comment]
You don't have to manually install if you had used `fastapi[all]`. Check `project.toml` for more understanding on optional dependencies.

Also, since `TestClient` is not needed for "normal" operations, `httpx` is optional as well. 

On a side note, it was clearly mentioned in release notes of both `FastAPI` and `starlette` about transitioning from `requests` to `httpx`

**Comment by pamelafox** [comment]
In terms of best practices, does it seem reasonable to put 'fastapi[all]' in requirements-dev.txt but just 'fastapi' in requirements.txt, so that additional libraries needed for testing are only pulled in development mode? That's my new approach.

**Comment by pythonweb2** [comment]
@pamelafox if you use a package manager like poetry, pdm, or pipenv, you can specify development dependencies so that you can exclude certain dependencies from production instances.

**Comment by ajferrario** [comment]
FWIW I'm not satisfied by the suggested answer here. When I install a package I should expect that package to install its own dependencies by default. If a more comprehensive installation statement is absolutely required then it should be included in the documentation.

**Reply by jgould22** [reply]
Httpx is installed by default unless you use the slim package.

**Comment by 4wuyan** [comment]
> When I install a package I should expect that package to install its own dependencies by default.

I agree with this.

Meanwhile, I think the current approach with fastapi is ok: it leaves the entry to some features open, while you get a message that you need to install extra things when you want to use that. (Feel like some in-app purchases 😂 ).

In a fresh install of fastapi==0.112.1, I get
```
>>> from fastapi.testclient import TestClient
Traceback (most recent call last):
  File ".../.venv/lib/python3.10/site-packages/starlette/testclient.py", line 33, in <module>
    import httpx
ModuleNotFoundError: No module named 'httpx'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File ".../.venv/lib/python3.10/site-packages/fastapi/testclient.py", line 1, in <module>
    from starlette.testclient import TestClient as TestClient  # noqa
  File ".../.venv/lib/python3.10/site-packages/starlette/testclient.py", line 35, in <module>
    raise RuntimeError(
RuntimeError: The starlette.testclient module requires the httpx package to be installed.
You can install this with:
    $ pip install httpx
```

So httpx is still not included by default.

However, for those who do not want to install `fastapi[all]`, https://github.com/fastapi/fastapi/pull/11503 introduced `fastapi[standard]`. You might give `fastapi[standard]` more consideration after https://github.com/fastapi/fastapi/pull/11935 makes it necessary if you'd like to use the fastapi CLI.
